codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs

```diff
@@ -2113,42 +2156,134 @@ typeRefsForExactName bhId reversedSegments = do
 --
 -- Get the list of type names for a given Reference within a given namespace.
-typeNamesForRefWithinNamespace :: BranchHashId -> NamespaceText -> Reference.TextReference -> Maybe ReversedSegments -> Transaction [ReversedSegments]
+-- Considers one level of dependencies, but not transitive dependencies.
+typeNamesForRefWithinNamespace :: BranchHashId -> PathSegments -> Reference.TextReference -> Maybe ReversedName -> Transaction [ReversedName]
 typeNamesForRefWithinNamespace bhId namespaceRoot ref maySuffix = do
   let namespaceGlob = toNamespaceGlob namespaceRoot
   let suffixGlob = case maySuffix of
         Just suffix -> toSuffixGlob suffix
         Nothing -> "*"
   queryListColCheck
     [sql|
-      SELECT reversed_name FROM scoped_type_name_lookup
-      WHERE root_branch_hash_id = :bhId
-            AND reference_builtin IS @ref AND reference_component_hash IS @ AND reference_component_index IS @
-            AND namespace GLOB :namespaceGlob
-            AND reversed_name GLOB :suffixGlob
-    |]
+        SELECT reversed_name FROM scoped_type_name_lookup
+        WHERE root_branch_hash_id = :bhId
+              AND reference_builtin IS @ref AND reference_component_hash IS @ AND reference_component_index IS @
+              AND namespace GLOB :namespaceGlob
+              AND reversed_name GLOB :suffixGlob
+        UNION ALL
+        SELECT (names.reversed_name || mount.reversed_mount_path) AS reversed_name
+        FROM name_lookup_mounts mount
+          INNER JOIN scoped_type_name_lookup names ON names.root_branch_hash_id = mount.mounted_root_branch_hash_id
+        WHERE mount.parent_root_branch_hash_id = :bhId
+              AND mount.mount_path GLOB :namespaceGlob
+              AND reference_builtin IS @ref AND reference_component_hash IS @ AND reference_component_index IS @
+              AND reversed_name GLOB :suffixGlob
+        |]
     \reversedNames -> for reversedNames reversedNameToReversedSegments
 
+-- | NOTE: requires that the codebase has an up-to-date name lookup index. As of writing, this
+-- is only true on Share.
+--
+-- Searches all dependencies transitively looking for the provided referent.
+-- Prefer 'termNamesForRefWithinNamespace' in most cases.
+-- This is slower and only necessary when resolving the name of references when you don't know which
+-- dependency it may exist in.
+--
+-- Searching transitive dependencies is exponential so we want to replace this with a more
+-- efficient approach as soon as possible.
+--
+-- Note: this returns the first name it finds by searching in order of:
+-- Names in the current namespace, then names in the current namespace's dependencies, then
+-- through the current namespace's dependencies' dependencies, etc.
+recursiveTermNameSearch :: BranchHashId -> Referent.TextReferent -> Transaction (Maybe ReversedName)
+recursiveTermNameSearch bhId ref = do
+  queryMaybeColCheck
+    [sql|
+        -- Recursive table containing all transitive deps
+        WITH RECURSIVE
```

Comment:
@tstat AFAICT this should be a breadth-first search that stops once it finds the first name but wouldn't mind if you could also stare at it for a bit 😄